src/vm/analysis/arithmetic_checker/mod.rs

```diff
@@ -0,0 +1,272 @@
+        use vm::functions::NativeFunctions::*;
+        match function {
+            FetchVar | GetBlockInfo | GetTokenBalance | GetAssetOwner | FetchEntry | SetEntry
+            | DeleteEntry | InsertEntry | SetVar | MintAsset | MintToken | TransferAsset
+            | TransferToken | ContractCall | StxTransfer | StxBurn | AtBlock | GetStxBalance => {
+                return Err(Error::FunctionNotPermitted(function));
+            }
+            Append | Concat | AsMaxLen | ContractOf | PrincipalOf | ListCons | Print
+            | AsContract => {
+                return Err(Error::FunctionNotPermitted(function));
+            }
+            Sha512 | Sha512Trunc256 | Secp256k1Recover | Secp256k1Verify | Hash160 | Sha256
+            | Keccak256 => {
+                return Err(Error::FunctionNotPermitted(function));
+            }
+            Add | Subtract | Divide | Multiply | CmpGeq | CmpLeq | CmpLess | CmpGreater
+            | Modulo | Power | Sqrti | Log2 | BitwiseXOR | And | Or | Not | Equals | If
+            | ConsSome | ConsOkay | ConsError | DefaultTo | UnwrapRet | UnwrapErrRet | IsOkay
+            | IsNone | Asserts | Unwrap | UnwrapErr | IsErr | IsSome | TryRet | ToUInt | ToInt
+            | Len | Begin | TupleMerge => self.check_all(args),
+            // we need to treat all the remaining functions specially, because these
+            //   do not eval all of their arguments (rather, one or more of their arguments
+            //   is a name)
+            TupleGet => {
+                // these functions use a name in the first argument
+                check_argument_count(2, args).map_err(|_| Error::UnexpectedContractStructure)?;
+                self.check_all(&args[1..])
+            }
+            Match => {
+                if !(args.len() == 4 || args.len() == 5) {
+                    return Err(Error::UnexpectedContractStructure);
+                }
+                // check the match input
+                self.check_expression(&args[0])?;
+                // check the 'ok' branch
+                self.check_expression(&args[2])?;
+                // check the 'err' branch
+                if args.len() == 4 {
+                    self.check_expression(&args[3])
+                } else {
+                    self.check_expression(&args[4])
+                }
+            }
+            Let => {
+                check_arguments_at_least(2, args)
+                    .map_err(|_| Error::UnexpectedContractStructure)?;
+
+                let binding_list = args[0]
+                    .match_list()
+                    .ok_or(Error::UnexpectedContractStructure)?;
+
+                for pair in binding_list.iter() {
+                    let pair_expression = pair
+                        .match_list()
+                        .ok_or(Error::UnexpectedContractStructure)?;
+                    if pair_expression.len() != 2 {
+                        return Err(Error::UnexpectedContractStructure);
+                    }
+
```

Comment:
Do we need to check that `pair_expression[0]` is an atom?

Reply:
No, we shouldn't need to here, because this check is invoked after the type checker (and that check is performed there).